portal-impl/src/com/liferay/portal/security/permission/ResourceActionsImpl.java

```diff
@@ -658,7 +661,7 @@ public void readAndCheck(ClassLoader classLoader, String... sources)
 		Set<String> portletNames = new HashSet<>();
 
 		for (String source : sources) {
-			_read(classLoader, source, portletNames);
+			_read(classLoader, source, rootElement -> _read(rootElement, null));
```

Comment:
This is wrong. Just looking at this method, no one populates `portletNames`, then the following for loop is dead code.

Reply:
@tinatian Ah this should be a typo in rebasing... I should pass "portletNames" as the second parameter of the `_read()` in the lambda.

When reading from a "source", **1)** the outter `_read` checks if the "source" points to other resource xml files (like default.xml in journal-services) or if "-ext" files exists, and **2)** call the inner `_read()` to read content from each file. This change decouples the two steps so caller can designate different content reading logic for 2) while reusing the logic of 1). I added this change in preparation of the next commit https://github.com/tinatian/liferay-portal/pull/2046/commits/d29089fb784a2a83e69ed8b1eef8262ede85a403 where I introduce a new method to only read the resource of the passed Portlet instance.

This commit is cherry-picked from the previous LPS-117531 (seperate portlet & model reading) pull.